### PR TITLE
Add support for static binary in GHA

### PR DESCRIPTION
**Pull request by lyrixx**

### CODE REVIEW

**Review by pyrech — APPROVED** [review]
That's an awesome feature to increase Castor adoption. Thanks :yellow_heart:

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,7 @@
 * Allow to listen to the symfony console events
 * Add a `compile` command that puts together a customizable PHP binary with a
   repacked castor app into one executable file
+* Distribute `castor.linux.amd64` automatically with the release
```
Comment (ruudk) [inline CHANGELOG.md:10]: Can we also add linux.arm64 and darwin.arm64?

Comment (lyrixx) [reply CHANGELOG.md:10]: nope :( Github does not support yet ARM64.
I asked to join [the beta](https://resources.github.com/devops/accelerate-your-cicd-with-arm-and-gpu-runners-in-github-actions/), but for now it's not possible

File: .github/workflows/artifacts.yml [hunk 2]
```diff
@@ -52,10 +54,72 @@ jobs:
...
+      - name: Upload the Linux static binary
+        uses: actions/upload-artifact@v4
+        with:
+          name: 'castor.darwin-amd64'
+          path: ./castor.darwin-amd64
+          if-no-files-found: error
+
   release:
-    name: Upload phars to the release
+    name: Upload artefacts to the release
```
Comment (tigitz) [inline .github/workflows/artifacts.yml:120]:
```suggestion
    name: Upload artifacts to the release
```

File: CHANGELOG.md (hunk 1 above)
Comment (ruudk) [reply CHANGELOG.md:10]: You can build for ARM using QEMU. It's just a bit slower, but I think it only runs on tag/release so might be fine?

Comment (lyrixx) [reply CHANGELOG.md:10]: I don't know how to do that :/ Do you want to give it a try? (but after this PR is merged, we are still not sure we want to merge it)

Comment (ruudk) [reply CHANGELOG.md:10]: Yes, let's have a look after this has been merged.

**Comment by tigitz** [comment]
A small note related to [my early comment](https://github.com/jolicode/castor/issues/257#issuecomment-1910180031)

The subset of extensions chosen to be compiled currently in the binary is really small, it was designed to run a simple "Hello World". 

My concern would be people trying to use it in their own real world Castor projects and hitting errors straight away. This would be a subpar experience.

Could we find a right balance of php extensions to include natively without bloating the binary too much ? 

Maybe we could take insipiration from the official docker image, running `docker run --rm php:8.3-alpine php -m
` gives :
```
[PHP Modules]
Core
ctype
curl
date
dom
fileinfo
filter
hash
iconv
json
libxml
mbstring
mysqlnd
openssl
pcre
PDO
pdo_sqlite
Phar
posix
random
readline
Reflection
session
SimpleXML
sodium
SPL
sqlite3
standard
tokenizer
xml
xmlreader
xmlwriter
zlib
```

Or maybe introducing a `slim` and normal version like docker official image tend to do would cover most use cases ?

**Comment by lyrixx** [comment]
ATM, the static binary weight <5mb, this is quite light IMHO.
What extension are missing?

**Comment by tigitz** [comment]
Compared to the official php docker image the missing extensions are:

- ctype
- curl
- dom
- fileinfo
- filter
- iconv
- mysqlnd
- openssl
- PDO
- pdo_sqlite
- readline
- session
- SimpleXML
- sodium
- sqlite3
- xml
- xmlreader
- xmlwriter

If we add all those extensions `bin/castor compile tools/phar/build/castor.linux-amd64.phar --php-extensions=mbstring,phar,posix,tokenizer,pcntl,ctype,curl,dom,fileinfo,filter,iconv,mysqlnd,openssl,pdo,pdo_sqlite,readline,session,simplexml,sodium,sqlite3,xml,xmlreader,xmlwriter` it takes 4 minutes on my laptop, including downloading, compared to 1 minutes 20 seconds before.

Size is `28M` vs `9MB` before.

But anyway, [I found those default libraries in docker image weren't here because they're popular but rather because they're non trivial to install](https://github.com/docker-library/php/pull/727#issuecomment-431503152)

So I guess Castor could define its own set of default extensions.

Personnaly I add in my own projects:
- ctype
- curl
- dom
- PDO
- pdo_pgsql
- openssl
- sodium

Which, aside from the pgsql requirements, would feel quite a good complementay default extension list to me.

**Comment by lyrixx** [comment]
> Size is `28M` vs `9MB` before.

something is strange on you computer. on github, it's only 4mb. [see this build](https://github.com/jolicode/castor/actions/runs/7980711446).

I'll made some bench with

```
ctype
curl
dom
PDO
pdo_pgsql
openssl
sodium
```

you really do some nice things with castor :D

**Comment by tigitz** [comment]
> something is strange on you computer. on github, it's only 4mb. [see this build](https://github.com/jolicode/castor/actions/runs/7980711446).

Not sure what's going on but I've tried from a freshly cloned project, phar is the same size from build but binary is still around 9M :man_shrugging: 

Maybe others experience the same issue ?

> you really do some nice things with castor :D

Yeah :sweat_smile: I manage backup scripts, send emails, scrape the latest update of a webpage, and have my own vault secret management based on symfony one. It's all about devops automation.

**Comment by lyrixx** [comment]
For the record, we had some concerns about security of the tool chain. 

We use https://github.com/dixyes/phpmicro and we didn't review all the code. But this packages is used a lot by Laravel (herd, native php, etc) 

So I think it's fine and I'll merge this PR since it's a real game changer !

**Pull request merged**